vmm/src/gdb.rs

```diff
@@ -0,0 +1,367 @@
+use std::{net::TcpListener, sync::mpsc};
+
+use gdbstub::{
+    arch::Arch,
+    target::{
+        ext::{
+            base::{
+                singlethread::{ResumeAction, SingleThreadOps, StopReason},
+                BaseOps, GdbInterrupt,
+            },
+            breakpoints::{Breakpoints, BreakpointsOps, HwBreakpoint, HwBreakpointOps},
+        },
+        Target, TargetError, TargetResult,
+    },
+    Connection,
+};
+#[cfg(target_arch = "x86_64")]
+use gdbstub_arch::x86::reg::X86_64CoreRegs as CoreRegs;
+#[cfg(target_arch = "x86_64")]
+use gdbstub_arch::x86::X86_64_SSE as GdbArch;
+use vm_memory::GuestAddress;
+
+#[cfg(target_arch = "x86_64")]
+type ArchUsize = u64;
+
+#[derive(Debug)]
+pub enum Error {
+    Vm(crate::vm::Error),
+    GdbRequest,
+    GdbResponseNotify(std::io::Error),
+    GdbResponse(mpsc::RecvError),
+    GdbResponseTimeout(mpsc::RecvTimeoutError),
+}
+type GdbResult<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum VmDebugStatus {
+    CommandComplete,
+    HitBreakPoint,
+}
+
+#[derive(Debug)]
+pub enum GdbResponsePayload {
+    Empty,
+    RegValues(Box<CoreRegs>),
+    MemoryRegion(Vec<u8>),
+    VmDebugStatus(VmDebugStatus),
+}
+
+pub type GdbResponse = std::result::Result<GdbResponsePayload, Error>;
+
+#[derive(Debug)]
+pub struct GdbRequest {
+    pub sender: mpsc::Sender<GdbResponse>,
+    pub payload: GdbRequestPayload,
+}
+
+#[derive(Debug)]
+pub enum GdbRequestPayload {
+    ReadRegs,
+    WriteRegs(Box<CoreRegs>),
+    ReadMem(vm_memory::GuestAddress, usize),
+    WriteMem(vm_memory::GuestAddress, Vec<u8>),
+    Pause,
+    Resume,
+    EnableSingleStep,
+    SetHwBreakPoint(Vec<vm_memory::GuestAddress>),
+}
+
+#[repr(u64)]
+#[derive(Debug)]
+pub enum GdbResponseEventKind {
+    ReadRegs = 1,
+    WriteRegs,
+    ReadMem,
+    WriteMem,
+    Pause,
+    Resume,
+    EnableSingleStep,
+    SetHwBreakPoint,
+}
+
+pub fn gdb_thread(mut gdbstub: GdbStub, port: u32) {
+    let addr = format!("0.0.0.0:{}", port);
```

Comment:
The listening address should be configurable.

Reply:
Does it have to be TCP? I wonder if a UNIX socket would be just as good (can always use socat if you want it on the netwwork)

Reply:
I didn't know that the newer version of GDB supports Unix domain sockets. It looks better than using TCP. I added the change as 0ed64ab.